Add GitHub workflow for `markdownlint`

its-miroma opened the pull request.

its-miroma committed:
Add GitHub workflow for `markdownlint`

its-miroma committed:
Add `MD054` rule

modmuss50 commented on .github/workflows/markdownlint.yml:
```diff
@@ -0,0 +1,51 @@
+
+jobs:
+  markdownlint:
+    runs-on: ubuntu-latest
```
Specify version

modmuss50 commented on .github/workflows/markdownlint.yml:
```diff
@@ -0,0 +1,51 @@
+
+      - id: titlecase
+        if: steps.node.outcome == 'success'
+        run: npm install markdownlint-rule-titlecase
```
Specify version

modmuss50 commented on .github/workflows/markdownlint.yml:
```diff
@@ -0,0 +1,51 @@
+        uses: DavidAnson/markdownlint-cli2-action@v16
+        with:
+          config: .markdownlint-cli2.yaml
+          globs: ${{ steps.changes.outputs.all_changed_files }}
```
Is there a good reason to only lint changed files? It would be easier to just do it all, and would also make any changes to the lint rules also apply to existing files.

its-miroma replied:
Performance reasons mainly, and also if all files are linted contributors may have to fix issues in unrelated files

A pull request shouldn't get blocked by this action because an unrelated file is not linted

its-miroma replied:
I guess that wouldn't be a problem if all files were linted to begin with, though

its-miroma replied:
> would also make any changes to the lint rules also apply to existing files

See #99, there should not be such problem after that's merged

its-miroma committed:
Add version numbers

IMB11 commented:
LGTM, #99 has been merged.

modmuss50 commented on .github/workflows/markdownlint.yml:
```diff
@@ -0,0 +1,51 @@
+            !versions/**
+
+      - id: node
+        if: steps.changes.outputs.any_changed == 'true'
```
I would just always lint the entire thing, we arent too bothered about speed here. Keep it simple.

IMB11 replied:
Some translations will be unlintable based on the way the translations are patched in by Crowdin, not nessecarily about speed. I agree that we should lint versioned pages if they're edited (just not translation pages from Crowdin)

its-miroma replied:
Does linting (aka modifying) translated files cause problems in Crowdin? That's my only concern

its-miroma replied:
> I would just always lint the entire thing

Even unchanged files? See <https://github.com/FabricMC/fabric-docs/pull/101#discussion_r1597707214>

IMB11 replied:
> Does linting (aka modifying) translated files cause problems in Crowdin? That's my only concern

No, its more the other way round, the linter would error out pretty much instantly with most of the non-latin languages (chinese, etc.), I had a try and it just doesn't support them very well.

IMB11 replied:
> Even unchanged files? See https://github.com/FabricMC/fabric-docs/pull/101#discussion_r1597707214

A good point was brought up though, if we change lint rules, it'd enforce us to change the old pages :+1:

IMB11 replied:
tldr:

- Lint all files.
- Ignore anything in versions/<version>/translated/
- Ignore anything in translated/

modmuss50 commented on .github/workflows/markdownlint.yml:
```diff
@@ -0,0 +1,51 @@
+
+      - id: titlecase
+        if: steps.node.outcome == 'success'
+        run: npm install markdownlint-rule-titlecase@v0.1.0
```
What does this do? is this magically picked up by the next step?

its-miroma replied:
To make sure that all headings use Title Case, I used a custom `markdownlint` rule.

> is this magically picked up by the next step?

See <https://github.com/its-miroma/fabric-docs/blob/e1c496ae37812f77bfa68cba9b4419af352c3af2/.markdownlint-cli2.yaml#L50-L51>

IMB11 reviewed (changes requested):
See comment by modmuss on linting everything.

its-miroma committed:
Simplify logic

- Only exclude `ŧranslated` files
- Only lint with `markdownlint-rule-titlecase`

its-miroma committed:
Lint all files, not only changed files

IMB11 commented:
This seems good, LGTM.

its-miroma committed:
Allow all HTML elements
